File: nodejs-express-js/src/callback.test.js
// The 3DS return is only as good as this checksum: it is what separates a payer coming back from
// the bank from somebody typing an order id into the address bar. The vectors below were computed
// outside this code, so a change in how the string is assembled shows up here as a failure rather
// than agreeing with itself.

import assert from 'node:assert/strict';
import test from 'node:test';
import { stubSettings } from './settings-stub.js';

stubSettings();
const { SIGNED_CALLBACK_FIELDS, validCallback } = await import('./callback.js');

// sha1('approved' + '1234567' + 'hf-abc' + 'test-merchant-control')
const CONTROL = '652ace404c4dfe8bba069ecee594ec23a89340e1';
const CALLBACK = {
  status: 'approved',
  orderid: '1234567',
  merchant_order: 'hf-abc',
  control: CONTROL,
};

test('a callback the gateway signed is accepted', () => {
  assert.equal(validCallback(CALLBACK), true);
});

test('every signed field is part of the checksum', () => {
  for (const field of ['status', 'orderid', 'merchant_order']) {
    assert.equal(validCallback({ ...CALLBACK, [field]: 'edited' }), false, `${field} is not signed`);
  }
});

test('a wrong control of the right length is rejected', () => {
  const wrong = `${CONTROL.slice(0, -1)}${CONTROL.endsWith('1') ? '2' : '1'}`;
  assert.equal(wrong.length, CONTROL.length);
  assert.equal(validCallback({ ...CALLBACK, control: wrong }), false);
});

test('a control of the wrong length is rejected, not thrown on', () => {
  // timingSafeEqual throws when the lengths differ, so this is the case that would surface as a
  // 500 rather than as a 403 if the length check were dropped.
  assert.equal(validCallback({ ...CALLBACK, control: 'short' }), false);
  assert.equal(validCallback({ ...CALLBACK, control: `${CONTROL}extra` }), false);
});

test('a request with nothing in it is rejected', () => {
  // sha1('' + '' + '' + 'test-merchant-control') — an empty callback still has a checksum, and it
  // is not the empty string, so a bare /result must not pass.
  assert.equal(validCallback({}), false);
  assert.equal(validCallback({ control: '1a66987ac24e927ff2979f83a41cb818936a9e62' }), true);
});

test('the signed fields travel in the order the page wants them', () => {
  assert.deepEqual(SIGNED_CALLBACK_FIELDS, ['status', 'orderid', 'merchant_order', 'control']);
});

File: go-react/callback_test.go
package main

import (
	"net/url"
	"testing"
)

// The 3DS return is only as good as this checksum: it is what separates a payer coming back from
// the bank from somebody typing an order id into the address bar. The vectors below were computed
// outside this code, and are the same ones the Express example uses, so neither can drift alone.

const testMerchantControl = "test-merchant-control"

// sha1("approved" + "1234567" + "hf-abc" + "test-merchant-control")
const testControl = "652ace404c4dfe8bba069ecee594ec23a89340e1"

func signedCallback() url.Values {
	return url.Values{
		"status":         {"approved"},
		"orderid":        {"1234567"},
		"merchant_order": {"hf-abc"},
		"control":        {testControl},
	}
}

func TestValidCallback(t *testing.T) {
	cfg.MerchantControl = testMerchantControl

	if !validCallback(signedCallback()) {
		t.Fatal("a callback the gateway signed was rejected")
	}

	// Every field the gateway signs has to be part of the checksum
	for _, field := range []string{"status", "orderid", "merchant_order"} {
		edited := signedCallback()
		edited.Set(field, "edited")
		if validCallback(edited) {
			t.Errorf("%s is not covered by the checksum", field)
		}
	}

	// A wrong control of the right length, which is what a guess looks like
	wrong := signedCallback()
	wrong.Set("control", testControl[:len(testControl)-1]+"0")
	if validCallback(wrong) {
		t.Error("a wrong control of the right length was accepted")
	}

	// A control of the wrong length must be rejected rather than panic
	for _, control := range []string{"", "short", testControl + "extra"} {
		odd := signedCallback()
		odd.Set("control", control)
		if validCallback(odd) {
			t.Errorf("control %q was accepted", control)
		}
	}
}

func TestEmptyCallbackStillHasAChecksum(t *testing.T) {
	cfg.MerchantControl = testMerchantControl

	// sha1("" + "" + "" + "test-merchant-control") — an empty callback is not the empty string,
	// so a bare /result must not pass on a missing control.
	if validCallback(url.Values{}) {
		t.Error("an empty callback with no control was accepted")
	}
	if !validCallback(url.Values{"control": {"1a66987ac24e927ff2979f83a41cb818936a9e62"}}) {
		t.Error("the checksum of an empty callback was rejected")
	}
}
